src/Microsoft.ML.Tokenizers/Model/Tiktoken.cs

```diff
@@ -0,0 +1,415 @@
+            }
+
+            if (_cache.Lookup(sequence, out int[] ids))
+            {
+                tokens = new(ids.Length);
+                tokens.Add(new Token(ids[0], sequence, (0, sequence.Length)));
+                for (int i = 1; i < ids.Length; i++)
+                {
+                    // One word split mapped to multiple Ids. Make the offset of the remaining token point at the end with zero width.
+                    tokens.Add(new Token(ids[i], "", (sequence.Length, sequence.Length)));
+                }
+
+                return tokens;
+            }
+
+            // cache miss
+            if (_vocab.TryGetValue(sequence, out int mappedId))
+            {
+                return new List<Token> { new(mappedId, sequence, (0, sequence.Length)) };
+            }
+
+            int[] encodedIds = BytePairEncoder.BytePairEncode(Encoding.UTF8.GetBytes(sequence), _encoder);
+            _cache.Add(sequence, encodedIds);
+
+            tokens = new List<Token>(encodedIds.Length);
+            tokens.Add(new Token(encodedIds[0], sequence, (0, sequence.Length)));
+            for (int i = 1; i < encodedIds.Length; i++)
+            {
+                // One word split mapped to multiple Ids. Make the offset of the remaining token point at the end with zero width.
+                tokens.Add(new Token(encodedIds[i], "", (sequence.Length, sequence.Length)));
+            }
+
+            return tokens;
+        }
+
+        /// <summary>
+        /// Tokenize a split sequence string to a list of Ids.
+        /// </summary>
+        /// <param name="sequence">The sequence to tokenize.</param>
+        /// <param name="isSpecialToken">Indicate if the token is a special token.</param>
+        /// <param name="accumulatedIds">The list of accumulated Ids.</param>
+        /// <returns>True if the operation succeeded, false otherwise.</returns>
+        public override bool TokenizeToIds(string sequence, bool isSpecialToken, List<int> accumulatedIds)
+        {
+            if (string.IsNullOrEmpty(sequence))
+            {
+                return true;
+            }
+
+            if (isSpecialToken)
+            {
+                if (_specialTokensEncoder is not null && _specialTokensEncoder.TryGetValue(sequence, out int id))
+                {
+                    accumulatedIds.Add(id);
+                    return true;
+                }
+
+                return false;
+            }
+
+            if (_cache.Lookup(sequence, out int[] tokenIds))
+            {
+                accumulatedIds.AddRange(tokenIds);
+                return true;
+            }
+
+            if (_vocab.TryGetValue(sequence, out int mappedId))
+            {
+                accumulatedIds.Add(mappedId);
+                return true;
+            }
+
+            int[] encodedIds = BytePairEncoder.BytePairEncode(Encoding.UTF8.GetBytes(sequence), _encoder);
+            _cache.Add(sequence, encodedIds);
```

Comment:
This cache grows unbounded?

Reply:
No, it is bounded and the bound can be controlled by the user. The default bound is 

LruCache.DefaultCacheSize = 8192;